crates/primitives/src/checkpoints.rs

```diff
@@ -121,6 +123,88 @@ pub struct StorageHashingCheckpoint {
     pub to: u64,
 }
 
+/// Saves the progress of a stage.
+#[main_codec]
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct StageCheckpoint {
+    /// The maximum block processed by the stage.
+    pub block_number: BlockNumber,
```

Comment:
this rises for regular stages and is falling for the header stage?

Reply:
not really, because otherwise it'll mess up the sync gap calculation. I planned to track it separately instead, using the `stage_checkpoint` field.